### ISSUE TITLE
NuGet package compatible with UWP?

### ISSUE THREAD

**Issue by nehmebilal** [autofac/Autofac.Extras.AttributeMetadata#1]
I am unable to install the `Autofac.Extras.AttributeMetadata` NuGet package in my UWP project.

Are there plans to support UWP?

Is there a work around other than resolving dependencies manually?

**Comment by tillig** [comment]
Currently the `Autofac.Extras.AttributeMetadata` package makes use of the `Autofac.Mef` package, which isn't compatible with UWP. There are not currently any plans to switch `Autofac.Mef` over to the `Microsoft.Composition` replacement for MEF. [This issue on Autofac.Mef has some additional info.](https://github.com/autofac/Autofac.Mef/issues/7)

**Comment by michael-hawker** [comment]
@tillig, Autofac still does properly import the Attributes off my classes though:

````
    [OperatorMetadata('%')]
    public class Mod : IOperation
    {
````

So, I can just use normal reflection to grab that attribute and inspect it.  My workaround for a similar structure (though without the benefit of lazy loading):

````
        IEnumerable<(IOperation Value, OperatorMetadataAttribute Metadata)> Operations { get; set; }

        public BasicCalculator(IEnumerable<IOperation> operators)
        {
            var oplist = new List<(IOperation Value, OperatorMetadataAttribute Metadata)>();
            // Load Operators & Retrieve Metadata
            foreach (var op in operators)
            {
                oplist.Add((op, 
                            op.GetType().GetTypeInfo().GetCustomAttribute(typeof(OperatorMetadataAttribute)) as OperatorMetadataAttribute)
                );
            }
            Operations = oplist;
        }
````

For comparison my structure was this before:

````
        IEnumerable<Lazy<IOperation, OperatorMetadataAttribute>> Operations { get; set; }

        public BasicCalculator(IEnumerable<Lazy<IOperation, OperatorMetadataAttribute>> operators)
        {
````

But at least none of my other code needed to change after that part.

However, couldn't Autofac just do this all for me automatically still?  
Or have a .WithAttributedMetadata<OperatorMetadataAttribute>() type helper I add at registration?

**Comment by tillig** [comment]
@michael-hawker When you say "Autofac still does properly import the Attributes off my classes though" the thing about your solution is that you're using reflection - a .NET mechanism, not an Autofac mechanism - to manually create metadata based on a custom attribute.

Could something entirely reflection based work? Sure. That has nothing to do with MEF.

However, if you look at [the actual code of how attribute metadata is tied to registrations](https://github.com/autofac/Autofac.Extras.AttributeMetadata/blob/develop/src/Autofac.Extras.AttributeMetadata/MetadataHelper.cs#L72) you'll see the whole point of it was to automatically search for attributes that _derive from `MetadataAttributeAttribute` from the MEF package_. MEF was responsible for providing the base attribute and a way for people to create metadata attributes that aren't strongly tied to Autofac.

MEF has changed a lot to support .NET Core and UWP. We don't have a plan to move any MEF support to .NET Core.

The stuff that wasn't MEF specific - `WithKeyAttribute`, `WithMetadataAttribute`, and the parameter filtering - all got moved to core Autofac and does support .NET Core / UWP. The stuff that is MEF specific - the scanning for MEF `MetadataAttributeAttribute` - was not moved.

The real key to replacing this would be to rewrite the whole thing to not use MEF (a huge breaking change, basically a whole new package) or rewrite it to use the updated version of MEF (also pretty breaking).

If you were to rewrite it without MEF, which is what your reflection-based solution basically does, you'd need to figure out a way to make the whole algorithm generically applicable. If someone wants to write their own metadata attribute, how would they do that? How would you tell Autofac which property/properties are interesting from a metadata standpoint? Is there a base attribute? Who owns that? Where does it live? Keeping in mind a lot of people don't want to have Autofac strongly bound to their classes, having a base attribute type (if there is one) that isn't an Autofac library provided thing is pretty important.

Long story short - the workaround for this is exactly what you already figured out: If you need attribute-based metadata, you can write your own extensions to do the type scanning for your own custom attributes.

If you do figure out how to make your mechanism generally applicable, you are 100% welcome to publish that as an extension package for Autofac and allow people to consume it.
